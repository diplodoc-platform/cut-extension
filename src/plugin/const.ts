export const ENV_FLAG_NAME = 'has-yfm-cut';

export const TokenType = {
    Cut: 'yfm_cut',
    CutOpen: 'yfm_cut_open',
    CutClose: 'yfm_cut_close',
    Title: 'yfm_cut_title',
    TitleOpen: 'yfm_cut_title_open',
    TitleClose: 'yfm_cut_title_close',
    Content: 'yfm_cut_content',
    ContentOpen: 'yfm_cut_content_open',
    ContentClose: 'yfm_cut_content_close',
} as const;

export const ClassNames = {
    Cut: 'yfm-cut',
    Title: 'yfm-cut-title',
    Content: 'yfm-cut-content',
} as const;
