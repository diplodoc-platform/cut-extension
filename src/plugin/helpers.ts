import type MdIt from 'markdown-it';

const CUT_REGEXP = /^{%\s*cut\s*["|'](.*)["|']\s*%}/;

export function matchOpenToken(tokens: MdIt.Token[], i: number) {
    return (
        tokens[i].type === 'paragraph_open' &&
        tokens[i + 1].type === 'inline' &&
        tokens[i + 1].content.match(CUT_REGEXP)
    );
}

export function matchCloseToken(tokens: MdIt.Token[], i: number) {
    return (
        tokens[i].type === 'paragraph_open' &&
        tokens[i + 1].type === 'inline' &&
        tokens[i + 1].content.trim() === '{% endcut %}'
    );
}

export function findCloseTokenIdx(tokens: MdIt.Token[], idx: number) {
    let level = 0;
    let i = idx;
    while (i < tokens.length) {
        if (matchOpenToken(tokens, i)) {
            level++;
        } else if (matchCloseToken(tokens, i)) {
            if (level === 0) {
                return i;
            }
            level--;
        }

        i++;
    }

    return null;
}
