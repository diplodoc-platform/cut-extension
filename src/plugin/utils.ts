export function hidden<B extends Record<string | symbol, unknown>, F extends string | symbol, V>(
    box: B,
    field: F,
    value: V,
) {
    if (!(field in box)) {
        Object.defineProperty(box, field, {
            enumerable: false,
            value: value,
        });
    }

    return box as B & Record<F, V>;
}

export type RawRuntime =
    | false
    | string
    | undefined
    | {
          script: string;
          style: string;
      };

export type Runtime =
    | {
          script: string;
          style: string;
      }
    | false;

export function getRuntime(runtime: RawRuntime) {
    if (typeof runtime === 'string') {
        return {
            script: runtime,
            style: runtime,
        };
    } else if (typeof runtime === 'undefined') {
        return {
            script: '_assets/cut-extension.js',
            style: '_assets/cut-extension.css',
        };
    } else {
        return runtime;
    }
}

declare const __dirname: string;
export function copyRuntime(
    {runtime, output}: {runtime: Runtime; output: string},
    cache: Set<string>,
) {
    if (!runtime) {
        return;
    }

    const PATH_TO_RUNTIME = '../runtime';
    const {join, resolve} = dynrequire('node:path');
    const runtimeFiles = {
        'index.js': runtime.script,
        'index.css': runtime.style,
    };
    for (const [originFile, outputFile] of Object.entries(runtimeFiles)) {
        const file = join(PATH_TO_RUNTIME, originFile);
        if (!cache.has(file)) {
            cache.add(file);
            copy(resolve(__dirname, file), join(output, outputFile));
        }
    }
}

export function copy(from: string, to: string) {
    const {mkdirSync, copyFileSync} = dynrequire('node:fs');
    const {dirname} = dynrequire('node:path');

    mkdirSync(dirname(to), {recursive: true});
    copyFileSync(from, to);
}

/*
 * Runtime require hidden for builders.
 * Used for nodejs api
 */
export function dynrequire(module: string) {
    // eslint-disable-next-line no-eval
    return eval(`require('${module}')`);
}
