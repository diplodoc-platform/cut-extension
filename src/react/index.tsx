export const CutRuntime = () => {
    return false;
};
