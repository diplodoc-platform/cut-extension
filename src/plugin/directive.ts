import type MarkdownIt from 'markdown-it';
import directives, {type MarkdownItWithDirectives} from 'markdown-it-directive';

import {ClassNames, ENV_FLAG_NAME, TokenType} from './const';

export const cutDirective: MarkdownIt.PluginSimple = (md) => {
    md.use(directives);

    (md as MarkdownItWithDirectives).blockDirectives['cut'] = ({
        state,
        content,
        contentStartLine,
        contentEndLine,
        inlineContent,
        inlineContentStart,
        inlineContentEnd,
        directiveStartLine,
        directiveEndLine,
    }) => {
        state.env ??= {};
        state.env[ENV_FLAG_NAME] = true;

        let token = state.push(TokenType.CutOpen, 'div', 1);
        token.block = true;
        token.attrSet('class', ClassNames.Cut);
        token.map = [directiveStartLine, directiveEndLine];

        token = state.push(TokenType.TitleOpen, 'div', 1);
        token.attrSet('class', ClassNames.Title);
        token.map = [inlineContentStart, inlineContentEnd];

        token = state.push('inline', '', 0);
        token.children = [];
        token.content = inlineContent || '';
        token.map = [directiveStartLine, directiveStartLine + 1];

        token = state.push(TokenType.TitleClose, 'div', -1);

        token = state.push(TokenType.ContentOpen, 'div', 1);
        token.attrSet('class', ClassNames.Content);
        token.map = [directiveStartLine + 1, directiveEndLine - 1];

        if (content) {
            const oldParent = state.parentType;
            const oldLineMax = state.lineMax;

            // @ts-expect-error bad types of state.parentType
            state.parentType = 'yfm-cut';
            state.line = contentStartLine!;
            state.lineMax = contentEndLine!;

            state.md.block.tokenize(state, contentStartLine!, contentEndLine!);

            state.lineMax = oldLineMax;
            state.parentType = oldParent;
        }

        token = state.push(TokenType.ContentClose, 'div', -1);

        token = state.push(TokenType.CutClose, 'div', -1);
        token.block = true;

        return true;
    };
};
