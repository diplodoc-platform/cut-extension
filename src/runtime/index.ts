import {GLOBAL_KEY} from './const';
import {YfmCutContoller} from './contoller';

import './styles/cut.scss';

if (typeof window !== 'undefined' && typeof document !== 'undefined' && !window[GLOBAL_KEY]) {
    window[GLOBAL_KEY] = new YfmCutContoller(document);
}

declare global {
    interface Window {
        [GLOBAL_KEY]: YfmCutContoller;
    }
}
