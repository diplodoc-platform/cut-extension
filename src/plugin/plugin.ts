import type MarkdownIt from 'markdown-it';
import type Core from 'markdown-it/lib/parser_core';

import {ClassNames, ENV_FLAG_NAME, TokenType} from './const';
import {findCloseTokenIdx, matchOpenToken} from './helpers';

export const cutPlugin: MarkdownIt.PluginSimple = (md) => {
    const rule: Core.RuleCore = (state) => {
        const tokens = state.tokens;
        let i = 0;

        while (i < tokens.length) {
            const match = matchOpenToken(tokens, i);

            if (match) {
                const closeTokenIdx = findCloseTokenIdx(tokens, i + 4);

                if (!closeTokenIdx) {
                    i += 3;
                    continue;
                }

                const newOpenToken = new state.Token(TokenType.CutOpen, 'details', 1);
                newOpenToken.attrSet('class', ClassNames.Cut);
                newOpenToken.map = tokens[i].map;

                const titleOpen = new state.Token(TokenType.TitleOpen, 'summary', 1);
                titleOpen.attrSet('class', ClassNames.Title);

                const titleInline = state.md.parseInline(
                    match[1] === undefined ? 'ad' : match[1],
                    state.env,
                )[0];

                const titleClose = new state.Token(TokenType.TitleClose, 'summary', -1);

                const contentOpen = new state.Token(TokenType.ContentOpen, 'div', 1);
                contentOpen.attrSet('class', ClassNames.Content);

                if (newOpenToken.map) {
                    const contentOpenStart = newOpenToken.map[0] + 1;
                    const contentOpenEnd = newOpenToken.map[0] + 2;

                    contentOpen.map = [contentOpenStart, contentOpenEnd];
                }

                const contentClose = new state.Token(TokenType.ContentClose, 'div', -1);

                const newCloseToken = new state.Token(TokenType.CutClose, 'details', -1);
                newCloseToken.map = tokens[closeTokenIdx].map;

                const insideTokens = [
                    newOpenToken,
                    titleOpen,
                    titleInline,
                    titleClose,
                    contentOpen,
                    ...tokens.slice(i + 3, closeTokenIdx),
                    contentClose,
                    newCloseToken,
                ];

                tokens.splice(i, closeTokenIdx - i + 3, ...insideTokens);

                state.env ??= {};
                state.env[ENV_FLAG_NAME] = true;

                i++;
            } else {
                i++;
            }
        }
    };

    try {
        md.core.ruler.before('curly_attributes', 'cut', rule);
    } catch (e) {
        md.core.ruler.push('cut', rule);
    }
};
