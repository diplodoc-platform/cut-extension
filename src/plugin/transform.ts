import type MarkdownIt from 'markdown-it';

import {cutPlugin} from './plugin';
import {type Runtime, copyRuntime, dynrequire, hidden} from './utils';
import {ENV_FLAG_NAME} from './const';
import {cutDirective} from './directive';

export type TransformOptions = {
    runtime?:
        | string
        | {
              script: string;
              style: string;
          };
    bundle?: boolean;
    /**
     * Enables directive syntax of yfm-cut.
     *
     * - 'disabled' – directive syntax is disabled.
     * - 'enabled' – directive syntax is enabled; old syntax is also enabled.
     * - 'only' – enabled only directive syntax; old syntax is disabled.
     *
     * @default 'disabled'
     */
    directiveSyntax?: 'disabled' | 'enabled' | 'only';
};

type NormalizedPluginOptions = Omit<TransformOptions, 'runtime' | 'directiveSyntax'> & {
    runtime: Runtime;
    directiveSyntax: NonNullable<TransformOptions['directiveSyntax']>;
};

const registerTransform = (
    md: MarkdownIt,
    {
        runtime,
        bundle,
        output,
        directiveSyntax,
    }: Pick<NormalizedPluginOptions, 'bundle' | 'runtime' | 'directiveSyntax'> & {
        output: string;
    },
) => {
    if (directiveSyntax === 'disabled' || directiveSyntax === 'enabled') {
        md.use(cutPlugin);
    }
    if (directiveSyntax === 'enabled' || directiveSyntax === 'only') {
        md.use(cutDirective);
    }
    md.core.ruler.push('yfm_cut_after', ({env}) => {
        hidden(env, 'bundled', new Set<string>());

        if (env?.[ENV_FLAG_NAME]) {
            env.meta = env.meta || {};
            env.meta.script = env.meta.script || [];
            env.meta.script.push(runtime.script);
            env.meta.style = env.meta.style || [];
            env.meta.style.push(runtime.style);

            if (bundle) {
                copyRuntime({runtime, output}, env.bundled);
            }
        }
    });
};

type InputOptions = {
    destRoot: string;
};

export function transform(options: Partial<TransformOptions> = {}) {
    const {bundle = true} = options;

    if (bundle && typeof options.runtime === 'string') {
        throw new TypeError('Option `runtime` should be record when `bundle` is enabled.');
    }

    const directiveSyntax = options.directiveSyntax || 'disabled';
    const runtime: Runtime =
        typeof options.runtime === 'string'
            ? {script: options.runtime, style: options.runtime}
            : options.runtime || {
                  script: '_assets/cut-extension.js',
                  style: '_assets/cut-extension.css',
              };

    const plugin: MarkdownIt.PluginWithOptions<{output?: string}> = function (
        md: MarkdownIt,
        {output = '.'} = {},
    ) {
        registerTransform(md, {
            directiveSyntax,
            runtime,
            bundle,
            output,
        });
    };

    Object.assign(plugin, {
        collect(input: string, {destRoot = '.'}: InputOptions) {
            const MdIt = dynrequire('markdown-it');
            const md = new MdIt().use((md: MarkdownIt) => {
                registerTransform(md, {
                    directiveSyntax,
                    runtime,
                    bundle,
                    output: destRoot,
                });
            });

            md.parse(input, {});
        },
    });

    return plugin;
}
