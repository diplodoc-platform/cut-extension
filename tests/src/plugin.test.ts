import transform from '@diplodoc/transform';
import dd from 'ts-dedent';
import MarkdownIt from 'markdown-it';

import * as cutExtension from '../../src/plugin';

const html = (text: string, opts?: cutExtension.TransformOptions) => {
    const {result} = transform(text, {
        plugins: [cutExtension.transform({bundle: false, ...opts})],
    });

    return result.html;
};

const meta = (text: string, opts?: cutExtension.TransformOptions) => {
    const {result} = transform(text, {
        plugins: [cutExtension.transform({bundle: false, ...opts})],
    });

    return result.meta;
};

const parse = (text: string, opts?: cutExtension.TransformOptions) => {
    const md = new MarkdownIt().use(cutExtension.transform({bundle: false, ...opts}));
    return md.parse(text, {});
};

describe('Cut extension - plugin', () => {
    describe('curly syntax', () => {
        it('should render simple cut', () => {
            expect(
                html(dd`
                {% cut "Cut title" %}

                Cut content

                {% endcut %}
                `),
            ).toMatchSnapshot();
        });

        it('should render simple cut with code in it', () => {
            expect(
                html(dd`
                {% cut "Cut title" %}

                \`\`\`
                Code
                \`\`\`

                {% endcut %}
                `),
            ).toMatchSnapshot();
        });

        it('should render siblings cuts', () => {
            expect(
                html(dd`
                {% cut "Cut title 1" %}

                Cut content 1

                {% endcut %}

                {% cut "Cut title 2" %}

                Cut content 2

                {% endcut %}
                `),
            ).toMatchSnapshot();
        });

        it('should render nested cuts', () => {
            expect(
                html(dd`
                {% cut "Outer title" %}

                Outer content

                {% cut "Inner title" %}

                Inner content

                {% endcut %}

                {% endcut %}
                `),
            ).toMatchSnapshot();
        });

        it('should render title with format', () => {
            expect(
                html(dd`
                {% cut "**Strong cut title**" %}

                Content we want to hide

                {% endcut %}
                `),
            ).toMatchSnapshot();
        });

        it('should close all tags correctly and insert two p tags', () => {
            expect(
                html(dd`
                * {% cut "Cut 1" %}

                  Some text

                  Some text

                  {% endcut %}
                `),
            ).toMatchSnapshot();
        });

        it('should close all tags correctly when given a bullet-list with several items', () => {
            expect(
                html(dd`
                * {% cut "Cut 1" %}

                  Some text

                  {% endcut %}

                * {% cut "Cut 2" %}

                  Some text

                  {% endcut %}

                * {% cut "Cut 3" %}

                  Some text

                  {% endcut %}
                `),
            ).toMatchSnapshot();
        });

        it('should dont add assets to meta if no yfm-cut is found', () => {
            expect(meta('paragraph')).toBeUndefined();
        });

        it('should add default assets to meta', () => {
            expect(
                meta(dd`
                {% cut "Cut title" %}

                Cut content

                {% endcut %}
                `),
            ).toStrictEqual({
                script: ['_assets/cut-extension.js'],
                style: ['_assets/cut-extension.css'],
            });
        });

        it('should add custom assets to meta', () => {
            expect(
                meta(
                    dd`
                    {% cut "Cut title" %}

                    Cut content

                    {% endcut %}
                    `,
                    {runtime: 'yfm-cut'},
                ),
            ).toStrictEqual({
                script: ['yfm-cut'],
                style: ['yfm-cut'],
            });
        });

        it('should add custom assets to meta 2', () => {
            expect(
                meta(
                    dd`
                {% cut "Cut title" %}

                Cut content

                {% endcut %}
                `,
                    {
                        runtime: {script: 'yfm-cut.script', style: 'yfm-cut.style'},
                    },
                ),
            ).toStrictEqual({script: ['yfm-cut.script'], style: ['yfm-cut.style']});
        });

        it('should parse markup with cut to token stream', () => {
            expect(
                parse(dd`


                {% cut "Cut _title_" %}

                Cut content

                {% endcut %}


                `),
            ).toMatchSnapshot();
        });
    });
});
