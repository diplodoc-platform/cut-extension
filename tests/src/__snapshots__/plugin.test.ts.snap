// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cut extension - plugin curly syntax should close all tags correctly and insert two p tags 1`] = `
<ul>
  <li>
    <details class="yfm-cut">
      <summary class="yfm-cut-title">
        Cut 1
      </summary>
      <div class="yfm-cut-content">
        <p>
          Some text
        </p>
        <p>
          Some text
        </p>
      </div>
    </details>
  </li>
</ul>
`;

exports[`Cut extension - plugin curly syntax should close all tags correctly when given a bullet-list with several items 1`] = `
<ul>
  <li>
    <details class="yfm-cut">
      <summary class="yfm-cut-title">
        Cut 1
      </summary>
      <div class="yfm-cut-content">
        <p>
          Some text
        </p>
      </div>
    </details>
  </li>
  <li>
    <details class="yfm-cut">
      <summary class="yfm-cut-title">
        Cut 2
      </summary>
      <div class="yfm-cut-content">
        <p>
          Some text
        </p>
      </div>
    </details>
  </li>
  <li>
    <details class="yfm-cut">
      <summary class="yfm-cut-title">
        Cut 3
      </summary>
      <div class="yfm-cut-content">
        <p>
          Some text
        </p>
      </div>
    </details>
  </li>
</ul>
`;

exports[`Cut extension - plugin curly syntax should render nested cuts 1`] = `
<details class="yfm-cut">
  <summary class="yfm-cut-title">
    Outer title
  </summary>
  <div class="yfm-cut-content">
    <p>
      Outer content
    </p>
    <details class="yfm-cut">
      <summary class="yfm-cut-title">
        Inner title
      </summary>
      <div class="yfm-cut-content">
        <p>
          Inner content
        </p>
      </div>
    </details>
  </div>
</details>
`;

exports[`Cut extension - plugin curly syntax should render siblings cuts 1`] = `
<details class="yfm-cut">
  <summary class="yfm-cut-title">
    Cut title 1
  </summary>
  <div class="yfm-cut-content">
    <p>
      Cut content 1
    </p>
  </div>
</details>
<details class="yfm-cut">
  <summary class="yfm-cut-title">
    Cut title 2
  </summary>
  <div class="yfm-cut-content">
    <p>
      Cut content 2
    </p>
  </div>
</details>
`;

exports[`Cut extension - plugin curly syntax should render simple cut 1`] = `
<details class="yfm-cut">
  <summary class="yfm-cut-title">
    Cut title
  </summary>
  <div class="yfm-cut-content">
    <p>
      Cut content
    </p>
  </div>
</details>
`;

exports[`Cut extension - plugin curly syntax should render simple cut with code in it 1`] = `
<details class="yfm-cut">
  <summary class="yfm-cut-title">
    Cut title
  </summary>
  <div class="yfm-cut-content">
    <pre>
      <code class="hljs">
        Code
      </code>
    </pre>
  </div>
</details>
`;

exports[`Cut extension - plugin curly syntax should render title with format 1`] = `
<details class="yfm-cut">
  <summary class="yfm-cut-title">
    <strong>
      Strong cut title
    </strong>
  </summary>
  <div class="yfm-cut-content">
    <p>
      Content we want to hide
    </p>
  </div>
</details>
`;
